contrib/babelfishpg_tds/src/backend/tds/tdsutils.c

```diff
@@ -541,3 +575,571 @@ TdsErrorContextCallback(void *arg)
+		return;
+
+	babelfish_db_id = get_database_oid(babelfish_db_name, true);
+	if (babelfish_db_id == target_db_id) /* rename active babelfish database */
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_IN_USE),
+				 errmsg("cannot rename active babelfish database")));
+}
+
+/*
+ *  handle_dropdb
+ *
+ *  Implements following two additional limitation to drop database stmt
+ *  1. block dropping an active babelfish database (indicated by babelfishpg_tsql.database_name)
+ *  2. block dropping an inactive babelfish database containing babelfish metadata (indicated by babelfishpg_tsql.enable_ownership_structure)
+ */
+static bool
+handle_dropdb(DropdbStmt *dropdb_stmt)
+{
+	Oid	        target_db_id = InvalidOid;
+	/*
+	 * Basic checkings to avoid non-privileged user to access metadata
+	 * allways let backend to handle error
+	 */
+	target_db_id = get_database_oid(dropdb_stmt->dbname, true);
+	if (target_db_id == InvalidOid)
+		return true;
+
+	/* Permission checks */
+	if (!pg_database_ownercheck(target_db_id, GetUserId()))
+		return true;
+
+	check_babelfish_dropdb_restrictions(target_db_id);
+	return true;
+}
+
+static void
+check_babelfish_dropdb_restrictions(Oid target_db_id)
+{
+	Relation    relsetting;
+	HeapTuple   tuple;
+	ScanKeyData scankey[2];
+	SysScanDesc scan;
+	const char  *babelfish_db_name = NULL;
+	Oid			babelfish_db_id = InvalidOid;
+	bool        has_bbf_md = false;
+
+	babelfish_db_name = GetConfigOption("babelfishpg_tsql.database_name", true, false);
+	if (!babelfish_db_name) /* not define */
+		return;
+	babelfish_db_id = get_database_oid(babelfish_db_name, true);
+	if (babelfish_db_id == target_db_id) /* drop active babelfish database */
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_IN_USE),
+				 errmsg("cannot drop active babelfish database")));
+
+	/*
+	 * check if it's an inactive babelfish database.
+	 * get db configs
+	 */
+	relsetting = table_open(DbRoleSettingRelationId, AccessShareLock);
+	ScanKeyInit(&scankey[0],
+				Anum_pg_db_role_setting_setdatabase,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target_db_id));
+	ScanKeyInit(&scankey[1],
+				Anum_pg_db_role_setting_setrole,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(InvalidOid));
+	scan = systable_beginscan(relsetting, DbRoleSettingDatidRolidIndexId, true,
+							NULL, 2, scankey);
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		bool        isnull;
+		Datum       datum;
+		ArrayType  	*configs = NULL;
+
+		datum = heap_getattr(tuple, Anum_pg_db_role_setting_setconfig,
+							RelationGetDescr(relsetting), &isnull);
+		if (!isnull)
+			configs = DatumGetArrayTypeP(datum);
+
+		if (configs && is_babelfish_ownership_enabled(configs))
+			has_bbf_md = true;
+	}
+	systable_endscan(scan);
+	table_close(relsetting, AccessShareLock);
+	if (has_bbf_md)
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_IN_USE),
+				 errmsg("babelfish metadata not removed, please run remove_babelfish before dropping database")));
```

Comment:
I'm not sure if this one is testable in automatic test without breaking the instance. If this cannot be tested in JDBC, please at least test this scenario locally and paste the record in PR/CR description.